proposals/0033-timely-vote-credits.md

```diff
@@ -0,0 +1,342 @@
+---
+simd: '0033'
+title: Timely Vote Credits
+authors:
+  - Bryan Ischo <bryan@ischo.com>
+category: Standard
+type: Core
+status: Draft
+created: 2023-01-30
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+Award a variable number of vote credits per voted on slot, with more credits
+being given for votes that have "less latency" than votes that have "more
+latency".
+
+
+## Motivation
+
+Vote credits are the accounting method used to determine what percentage of
+inflation rewards a validator earns on behalf of its stakers.  Currently, when
+a slot that a validator has previously voted on is "rooted", it earns 1 vote
+credit.  A "rooted" slot is one which has received full committment by the
+validator (i.e. has been finalized).
+
+One problem with this simple accounting method is that it awards one credit
+per rooted slot regardless of when it was voted on.  This means that a
+validator can delay its voting for many slots in order to survey forks and
+make votes that are more likely to be rooted.  This strategy saves fees on
+extraneous votes, and minimizes no voting periods due to lockout.
+
+This is not just a theoretical concern; a number of validators appear to be
+using this technique to inflate their vote credits, some having been doing so
+for over a year.  The number of validators using this approach is increasing
+over time.  It is estimated that about 7.5 million stake, or roughly 2% of all
+stake on the network, is using an intentional vote lagging strategy.
+
+## Alternatives Considered
+
+It's hard to conceive an alternative to this proposal since the problem can
+only be addressed via economic incentive to reduce vote lagging.  The only
+alternatives would be in the specific characteristics of how credits are
+awarded based on vote latency; several parameterizations were considered and
+this was written about in an early description of this issue:
+
+https://www.shinobi-systems.com/timely_voting_proposal
+
+This website is quite out of date however; the parameterizations that were
+proposed there ("4 16 1", see later in this document for a description of what
+these numbers mean) has been long since superceded by different values ("2 8
+1").  This is because the overall performance of the Solana mainnet cluster
+has matured to a degree where "tighter timings" are tolerable.
+
+## New Terminology
+
+**Vote Latency** is defined as the difference, in slots, between the slot
+being voted on within a vote transaction, and the slot in which the vote
+transaction itself lands.  The minimum possible vote latency for any voted-on
```

Comment:
> and the slot in which the vote transaction itself lands

I have some concern about this w.r.t. the proposed grace period. Currently, the leader scheduler gives leaders 4 slots at a time to be leader. The leader chooses which vote transactions get packed into those blocks.

As a malicious operator, I now have some incentive to not pack competing validator's votes; by delaying them, it reduces their rewards, and makes my validator appear more attractive to potential stakers.

I think with the current system, there's no real incentive to do this because it doesn't matter if I don't pack them, the next leader will and they'll get the same reward. And now I've only lost out on the transaction fees myself. 

I wonder if the grace period should be at least 4 slots to avoid this? Even so, I can imagine delaying processing votes until my 4th slot for the same reason while still earning the fees.

Haven't though through it entirely, but a potential way to address this is by making vote credits based on slot latency mod 4 (subject to change if leader schedule does).

Reply:
In doing so validators will be taking a risk of loss of block rewards: deferring any transactions until a later block will miss out on block rewards if those later blocks are skipped.  In addition, deferring votes to later blocks increases the chances that those blocks will be skipped as votes which could have landed earlier to "strengthen" the fork will land later instead, giving greater chance for a competing fork to win.

Both of these risks are anti-incentive for the validator to intentionally defer votes.

The effect that a validator can have on its "competitors" by deferring votes is very minor: a validator only gets approximately its stake weight in slots per epoch.  I don't think this has ever risen much above 3% since the early days of mainnet-beta, and anything above 2% is pretty rare.

Practically speaking this means that a validator could at worst reduce about 2% of a competitor's votes by 2 slots, or assuming an average of 7.5 credits per slot, a reduction from 7.5 to 5.5 credits for 2% of the target validators slots.  This is a reduction in 0.5% of credits -- and at the cost of greater risk of lost block rewards to the offending validator.

This issue was discussed in other forums previously and the conclusion was that it's not an issue large enough to counterbalance the benefits of the proposal.

But I'd be happy to hear your further thoughts on this.

Reply:
> In addition, deferring votes to later blocks increases the chances that those blocks will be skipped as votes which could have landed earlier to "strengthen" the fork will land later instead, giving greater chance for a competing fork to win.

Adding to this: 

Votes for blocks on different forks cannot land (they will fail as invalid tx in vote program), so a leader can only censor votes for their own fork - which as zan mentioned hurts their chance of being selected in the heaviest fork calculation. I imagine a competing validator would have a sizable stake so this disincentives such behavior. 

At best a leader could censor tiny validators which would have a minimal reward for loss of tx fees. Based on the numbers zan cited this seems unlikely to be beneficial. 

I think if any such adversarial block packing strategy emerges successfully the best way to counteract it would be to tweak the ratio of block rewards/vote tx fees/vote credits